File: src/mycli/domain/runtime/shell_lifecycle.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


ShellLifecycleKind = Literal[
    "shell.started",
    "shell.output",
    "shell.completed",
    "shell.removed",
    "shell.list.updated",
]


@dataclass(frozen=True, slots=True)
class ShellLifecycleEvent:
    kind: ShellLifecycleKind
    shell_id: str
    owner_session_id: str
    call_id: str | None
    sequence: int
    command_preview: str
    background: bool
    process_state: str
    terminal_state: str | None = None
    exit_code: int | None = None
    output_delta: str = ""
    next_cursor: int = 0
    output_chars: int = 0
    omitted_output_chars: int = 0
    cleanup_result: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    active_background_count: int | None = None
    shell_kind: str | None = None
    shell_edition: str | None = None

    def to_tui_payload(self) -> dict[str, object]:
        return {
            "shell_id": self.shell_id,
            "call_id": self.call_id,
            "sequence": self.sequence,
            "command_preview": self.command_preview,
            "background": self.background,
            "process_state": self.process_state,
            "terminal_state": self.terminal_state,
            "exit_code": self.exit_code,
            "output_delta": self.output_delta,
            "next_cursor": self.next_cursor,
            "output_chars": self.output_chars,
            "omitted_output_chars": self.omitted_output_chars,
            "cleanup_result": self.cleanup_result,
            "started_at": self.started_at,
            "completed_at": self.completed_at,
            "active_background_count": self.active_background_count,
            "shell_kind": self.shell_kind,
            "shell_edition": self.shell_edition,
        }


__all__ = ["ShellLifecycleEvent", "ShellLifecycleKind"]

File: backend/packages/core/src/shell-lifecycle.ts
export const SHELL_LIFECYCLE_OUTPUT_CHUNK_MAX_CHARS = 16_384;

export type ShellLifecycleKind =
	| "shell.started"
	| "shell.output"
	| "shell.completed"
	| "shell.removed"
	| "shell.list.updated";

export type ShellTransportKind = "pipe" | "unix_pty" | "windows_conpty";

export interface ShellLifecycleEvent {
	readonly type: "shell_lifecycle";
	readonly kind: ShellLifecycleKind;
	readonly shellId: string;
	readonly ownerSessionId: string;
	readonly callId: string | null;
	readonly sequence: number;
	readonly commandPreview: string;
	readonly background: boolean;
	readonly processState: string;
	readonly transport?: ShellTransportKind;
	readonly tty: boolean;
	readonly yielded: boolean;
	readonly terminalState?: string;
	readonly exitCode?: number;
	readonly outputDelta?: string;
	readonly nextCursor?: number;
	readonly outputChars?: number;
	readonly omittedOutputChars?: number;
	readonly cleanupResult?: string;
	readonly startedAt?: string;
	readonly completedAt?: string;
	readonly activeBackgroundCount?: number;
	readonly shellKind?: string;
	readonly shellEdition?: string;
}
